scheme.py

```diff
@@ -1,62 +1,187 @@
+from __future__ import annotations
+
 from dataclasses import dataclass
-from typing import List, Optional, Tuple, Iterable
+from typing import (
+    List, Optional, Tuple, Iterable, Sequence, Union, cast, TYPE_CHECKING)
+
+if TYPE_CHECKING:
+    from environment import Environment
 
 
-@dataclass
```

Comment:
For SExp, I figured we don't need the methods that dataclass generates, so rather than manually disabling them all I just wrote an explicit constructor.
I removed frozen from the others because I needed to set the environments.